### ISSUE TITLE
SQLite times out during query execution when using `Promise.all()` / concurrent

### ISSUE THREAD

**Issue by macjuul** [prisma/prisma#9562]
### Bug description

I'm attempting to await multiple queries at once using Promise.all(), something which works fine on MySQL, however on SQLite this gives the following error

```
touch(): Error: 
Invalid `prisma.session.update()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: ConnectionError(Timed out during query execution.) })
Error: 
Invalid `prisma.session.update()` invocation:


  Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: ConnectionError(Timed out during query execution.) })
    at cb (xxx\node_modules\@prisma\client\runtime\index.js:36956:17)
```

I'm hoping to use Promise.all() to execute and await multiple queries at once as performance optimization. I have also found the library [prisma-session-store](https://www.npmjs.com/package/@quixo3/prisma-session-store) to use a similar pattern thus also resulting in this error.

It is also worth mentioning the error isn't always thrown. It seems fairly inconsistent when the error is and isn't thrown, however most of the time it is.

### How to reproduce

1. Pass multiple concurrent queries into Promise.all()
2. Error is usually thrown

### Expected behavior

The queries complete concurrently and the Promise.all() resolves. This is the behavior in MySQL.

### Prisma information

Example:

```typescript
await Promise.all([
	database.tableOne.update({ ...your query... }),
	database.tableTwo.update({ ...your query... })
]);
```

### Environment & setup

- OS: Windows
- Database: SQLite
- Node.js version: v14.17.1

### Prisma Version

```
prisma                : 2.30.3
@prisma/client        : 3.1.1
Current platform      : windows
Query Engine (Binary) : query-engine b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules\@quixo3\prisma-session-store\node_modules\@prisma\engines\query-engine-windows.exe)
Migration Engine      : migration-engine-cli b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules\@quixo3\prisma-session-store\node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine  : introspection-core b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules\@quixo3\prisma-session-store\node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary         : prisma-fmt b8c35d44de987a9691890b3ddf3e2e7effb9bf20 (at node_modules\@quixo3\prisma-session-store\node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash  : b8c35d44de987a9691890b3ddf3e2e7effb9bf20
Studio                : 0.423.0
```

**Comment by alecho** [comment]
I'm seeing this as well.

**Comment by janpio** [comment]
Can one of you help us by providing a simple to use reproduction? The faster we get to running code that causes the same errors you are showing here, the better. Thanks.

**Comment by alecho** [comment]
> Can one of you help us by providing a simple to use reproduction? The faster we get to running code that causes the same errors you are showing here, the better. Thanks.

I should have some free cycles coming up.I think this is a SQLite issue and possibly related to #10306 . I think it's likely due to something. Looks like an a solution was proposed there so I'll check out that code to see if it helps before I create a repo that should reproduce this and like it here.

**Comment by HartS** [comment]
I'm seeing the same issue using prisma with sqlite, unrelated to seeding or Promise.all();

When executing 5 updates in parallel (not even on the same records, just the same table), it seems to occur with just one of the records. In my case I'm running the updates in parallel by using forEach over a set of records I need to update, but I'm going to work around it by using a for loop with an await inside.

**Comment by regs79** [comment]
> I'm seeing the same issue using prisma with sqlite, unrelated to seeding or Promise.all();
> 
> When executing 5 updates in parallel (not even on the same records, just the same table), it seems to occur with just one of the records. In my case I'm running the updates in parallel by using forEach over a set of records I need to update, but I'm going to work around it by using a for loop with an await inside.

I had the same issue, updating a handful of records with sqlite both with and without Promise.all(). Switching to the for loop with await inside did the trick.

**Comment by othmanhammouche** [comment]
> > I'm seeing the same issue using prisma with sqlite, unrelated to seeding or Promise.all();
> > When executing 5 updates in parallel (not even on the same records, just the same table), it seems to occur with just one of the records. In my case I'm running the updates in parallel by using forEach over a set of records I need to update, but I'm going to work around it by using a for loop with an await inside.
> 
> I had the same issue, updating a handful of records with sqlite both with and without Promise.all(). Switching to the for loop with await inside did the trick.

for loop did the trick for me too here

**Comment by aqrln** [comment]
Related: https://github.com/prisma/prisma/issues/10403

**Comment by MDrollette** [comment]
adding `?connection_limit=1` to the connection string also solves the issue of locking errors causing instant failures. Depending on your use-case you might get fewer errors from multiple concurrent requests when using a single connection - whereas replacing Promise.all with a for loop only helps for that individual request. I suspect when there's more than 1 connection being used any lock error fails instantly, when it should instead wait and retry (up to `socket_timeout`)

**Comment by janpio** [comment]
Note: there are multiple issues about this now all grouped under https://github.com/prisma/prisma/issues?q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+Timed+out+during+query+execution%22 One of them is https://github.com/prisma/prisma/issues/10403 which points out that our error message here is currently highly misleading and should be changed/fixed.

**Comment by noxware** [comment]
I'm not using `Promise.all` but I have a CLI app that runs `n` worker functions concurrently and I've been fighting this problem all night.

I've written a function serializer and wrapped all prisma calls with it and it worked.
But then I read the `?connection_limit=1` trick which worked without my serialization stuff, which is better.

I've started using Prisma a month ago and I feel like SQLite support is in general very poor. It's very demotivating. 

Extra note: Ensure that you don't have zombie node.js process in background accessing to the database or the `connection_limit` trick will not work 😅 I'm saying this because after hours of trying every possible fix and configuration, the final problem was that I accidently had 3 zombie nodejs processes in the background touching the db.

**Comment by noxware** [comment]
Now that I mentioned the zombie processes problem... Can Prisma Studio be interfering too if run it alongside my app?

**Comment by Nedi11** [comment]
bump

**Comment by Jolg42** [comment]
I found out, after reproducing this error from a similar report https://github.com/prisma/prisma/issues/11789 with concurrent queries in https://github.com/prisma/prisma/pull/22723

That `?connection_limit=1` to the `url` made this error disappear -> The queries were failing and are successful with the added parameter.

Example
```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = "file:./dev.db?connection_limit=1"
}
```

If you try that, let us know if that works for you, or not, we're curious to hear from you 🙌🏼

**Comment by msurdi** [comment]
I think this workarounds the problem when only accessing sqlite from a single process. If you access SQLite from more than one process (for example, a worker) then the problem shows up again.

I was having this problem when using Prisma a few months ago. I’ve moved the project to Knex and the problem went away. When a writer is making an operation the second writer would just block and wait until it can proceed. I think this connection limit is working on a single process because it’s queuing the writes within prisma, not within SQLite. Well, I don’t know Prisma internals so I’m speculating here.


TL;DR you should be able to write to a SQLite database from multiple processes, and you can’t with Prisma.


Hope this gets resolved. Thanks!

**Comment by Jolg42** [comment]
@msurdi same thinking here, thanks for writing down your thoughts 💚

**Comment by zaquas77** [comment]
Hi @msurdi and @Jolg42 

I have the same problem with a minimal test

be:
```
import express from 'express';
import { PrismaClient } from '@prisma/client';
import cors from 'cors';

const host = process.env.HOST ?? 'localhost';
const port = process.env.PORT ? Number(process.env.PORT) : 3000;

const app = express();
app.use(cors());

app.put('/:id', async (req, res) => {
  let user: unknown;
  const prisma = new PrismaClient();
  try {
    user = await prisma.user.update({
      where: { id: Number(req.params.id) },
      data: { name: `${crypto.randomUUID()}` },
    });
  } catch (error) {
    console.log(error);
  } finally {
    await prisma.$disconnect();
  }
  res.json(user);
});

app.listen(port, host, () => {
  console.log(`[ ready ] http://${host}:${port}`);
})
```
my `.env` file is
```
DATABASE_URL="file:./dev.db?connection_limit=1"
```

and a angular component for fe:
```
import { Component, inject } from '@angular/core';
import { firstValueFrom } from 'rxjs';
import { HttpClient } from '@angular/common/http';

@Component({
  standalone: true,
  selector: 'krug-root',
  template: '<button (click)="run()">run</button>'
})
export class AppComponent {
  title = 'krug';

  _http = inject(HttpClient);

  run = async () => {
    const promises = [];
    for (let index = 0; index < 100; index++) {
      promises.push(firstValueFrom(this._http.put('http://localhost:3000/1', {})));
    }
    const returns = await Promise.all(promises);
    console.log(returns);
  };
}
```
In the be console I read this errors:
```
PrismaClientUnknownRequestError: 
Invalid `prisma.user.update()` invocation in
/Users/krug/apps/dexs/src/main.ts:15:30

  12 let user: unknown;
  13 const prisma = new PrismaClient();
  14 try {
→ 15   user = await prisma.user.update(
Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: ConnectionError(Timed out during query execution.), transient: false })
    at si.handleRequestError (/Users/krug/node_modules/@prisma/client/runtime/library.js:125:7007)
    at si.handleAndLogRequestError (/Users/krug/node_modules/@prisma/client/runtime/library.js:125:6151)
    at si.request (/Users/krug/node_modules/@prisma/client/runtime/library.js:125:5859)
    at l (/Users/krug/node_modules/@prisma/client/runtime/library.js:130:9805)
    at /Users/krug/apps/dexs/src/main.ts:15:12 {
  clientVersion: '5.8.1'
}
```
I hope this can help someone to fix the problem 💚

**Comment by 4lejandrito** [comment]
Hi guys, the only way I am able to workaround this issue is by implementing retries:

```
prisma.$use(async (params, next) => {
  let attempts = 0
  while (attempts < 3) {
    try {
      return await next(params)
    } catch (error: any) {
      if (error.message.includes('ConnectionError')) {
        console.log(
          `Attempt ${attempts + 1}: Retrying after error: ${error.message}`
        )
        attempts++
        await delay(1000)
      } else {
        throw error
      }
    }
  }
  throw `No success after ${attempts + 1}`
})
```
With this code whenever the locking occurs it will retry and so far it has worked every time.

**Comment by jkomyno** [comment]
Hi @macjuul, @msurdi, @4lejandrito, @zaquas77 and others, I know this is an old issue, but we believe we now have a fix.

We plan on releasing it with the next stable version of Prisma (`5.16.0`), but in the meantime, we'd like to collect some early feedback directly from the impacted users. Can you please try installing Prisma `5.16.0-integration-sqlite-fix.1`, and confirm that you can no longer experience this bug with it? Thanks!

```sh
pnpm install -D prisma@5.16.0-integration-sqlite-fix.1
pnpm install @prisma/client@5.16.0-integration-sqlite-fix.1
```

(Please note that this Prisma version is __unstable__, so we don't recommend running it in production.)

**Comment by 4lejandrito** [comment]
Hey I am trying but I get the same errors. I updated prisma like this:
```
"prisma": "5.16.0-integration-sqlite-fix.1",
```
But I still have `"@prisma/client": "5.7.1",`. I suspect I am still using old code. How can I update prisma/client to your new version? if I put `"@prisma/client": "5.16.0-integration-sqlite-fix.1",` it fails to install.

Regards!

**Comment by jkomyno** [comment]
@4lejandrito It's also important to re-generate the Prisma Client (whose version needs to be `5.16.0-integration-sqlite-fix.1`) via `prisma generate`. I've updated my comment to highlight that :)

**Comment by 4lejandrito** [comment]
I do not experience the issue anymore with that version. Thanks so much guys!

**Comment by zaquas77** [comment]
Hi   @jkomyno,
I tried with this new version (5.16.0-integration-sqlite-fix.1) and worked like a charm! thanks very much for all your efforts

**Comment by jkomyno** [comment]
Hello! This issue was solved via https://github.com/prisma/prisma-engines/pull/4907. The fix is already available on Prisma 5.15.1.

Please make sure you update the `prisma` CLI, the `@prisma/client` library, and run `prisma generate` again :)
